core/src/main/java/org/elasticsearch/ingest/PipelineStore.java

```diff
@@ -154,18 +115,58 @@ public void close() throws IOException {
         IOUtils.close(closeables);
     }
 
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        IngestMetadata ingestMetadata = event.state().getMetaData().custom(IngestMetadata.TYPE);
+        if (ingestMetadata == null) {
+            return;
+        }
+
+        Map<String, Pipeline> pipelines = new HashMap<>();
+        for (Map.Entry<String, Map<String, Object>> entry : ingestMetadata.getPipelines().entrySet()) {
+            String id = entry.getKey();
+            try {
+                @SuppressWarnings("unchecked")
+                Map<String, Object> pipelineConfigCopy = (Map<String, Object>) deepCopy(entry.getValue());
+                pipelines.put(id, constructPipeline(id, pipelineConfigCopy));
```

Comment:
Looked at it and it makes it messy if we have to pass this down all the way down to ConfigurationUtils. 

Maybe instead we should not store the config a map of maps, but as bytes reference. After the pipeline has been created we don't use it any more.